### ISSUE TITLE
Keycard NFC tag remove Issue: Java Security Exception

### ISSUE THREAD

**Issue by localh0rs7** [status-im/status-react#14815]
## Problem 
Pairing the Keycard on Android 12 in Status.im (`Device 1`) causes problems.
Since some time Android 12 seems to have problem with the NFC connectivity when pairing and connecting the keycard in Status as well as signing transactions.

#### Expected behavior
When pairing a fresh keycard with the newest applet my Status App in Android 12, the app should go through the pairing process, requesting the pairing password and give access to the wallet.

#### Actual behavior
The following Java Runtime Error occurs when pairing a new card or connecting an existing card:

`AndroidRuntime: java.lang.SecurityException: Permission Denial: Tag ( ID: XX XX XX XX XX XX XX ) is out of date`

So this leads me to the conclusion that something happens during the NFC Rediscover routine that is not suppose to happen (maybe due to an uncatched exception when a Nfc reconnect happens in Android 12 when the card is not placed correctly on the sensor, which happens a lot during discovery).

### Reproduction

1) Open Status
2) Pair an existing keycard or connect an already paired keycard to the status app on Android 13. 
3.) When scanning the NFC Card, the following Java Exception Occurs:

 `AndroidRuntime: java.lang.SecurityException: Permission Denial: Tag ( ID: XX XX XX XX XX XX XX ) is out of date.` 

The app should crash after that exception.

4.) Other errors (APDU failures on `Device 2`) are also observed if connecting the card to an Android 11 after step 1 & 2.

`AndroidRuntime: java.lang.IllegalArgumentException: APDU response must be at least 2 bytes`

The pairing and connecting of the Keycard in Android 10 however still works.

### Additional Information
`Device 1:`
- Status version: 1.20
- Operating System: Android 12 (CalyxOS  4.5.0)
- Device: Fairphone 4 (PF4)

`Device 2:`
- Status version: 1.20
- Operating System: Android 11
- Device: Samsung Galaxy A50

After some investigation this seems to be a known issue at AOSP:

https://issuetracker.google.com/issues/238257723

It seems that this SecurityException was introduced in AOSP commit [113f3a4]
It is thrown when an NFC tag is removed.

The relevant Diff is given here:
https://android.googlesource.com/platform/frameworks/base/+/113f3a4a1d9171db7456aa9ae8b0a2b50545c326%5E%21/

**A history of the relevant code and the original Android 12 upstream issue at is attached below:** - NFC Tag Routines in the Android Nfc Service:

Methods: 
`void setTagUpToDate(long cookie)`
`boolean isTagUpToDate(long cookie)`

The return values from this functions should be checked as well as Catched Exceptions on tag removal.

https://android.googlesource.com/platform/packages/apps/Nfc/+/master/src/com/android/nfc/NfcService.java#2124
https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/nfc/Tag.java#380

An example fix for reference:
https://github.com/open-keychain/open-keychain/pull/2804

**Comment by flexsurfer** [comment]
cc @bitgamma

**Comment by Menduist** [comment]
Same issue for me

**Comment by localh0rs7** [comment]
@bitgamma 
@guylouis 
@flexsurfer 

If someone could point me into the right direction, I can also try to fix the issue by myself:

Steps I already did:
- Compiling the Closure Code of the Status Source to JavaScript
- Inspecting the pairing and keycard connect functions in `status-mobile/app`:
`status_im.keycard.real_keycard.js` ,
`status_im.keycard.login.js` :
    - Functions: 
    - `status_im.keycard.real_keycard.pair(){}`
    -  `status_im$keycard$login$login_with_keycard(var_args){}`
    - eventually also some signing parts in `status_im.keycard.sign.js` that use the keycard to authenticate.
    
    Is this the right trace to try to add the catch{} instruction or is there another higher level application routine that can handle the `java.lang.SecurityException`?
    Where are the above `TagUpToDate() `Methods called?

**Comment by flexsurfer** [comment]
hey @localh0rs7 , @bitgamma is off, when he's back he'll take a look, thanks

**Comment by bitgamma** [comment]
I am back from my holidays, I will have a look at this soon

**Comment by churik** [comment]
@localh0rs7 Hello!
Would you mind checking whether the issue is reproducible for you on https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-230206-124925-0fa59b-pr14997-universal.apk ?

We are unable to replicate the issue on Samsung Galaxy A52.
![IMAGE 2023-02-20 11:27:56](https://user-images.githubusercontent.com/4557972/220080007-03a650e6-ed0b-4994-bf54-a27870ac70b4.jpg)

**Comment by localh0rs7** [comment]
For me I cannot test the whole thing on my Samsung Galaxy A52 because it still runs with Android 11 and I am not sure if I will ever get any newer with the official ROM.

I tested it on Samsung A52 with Android 11 and it still works (like before). 
Only the following other exception seems to get raised (sometimes - only in negligible amount of tries and only on the Samsung Device with Android 11):
`AndroidRuntime: java.lang.IllegalArgumentException: APDU response must be at least 2 bytes`

On my Fairphone 4 (Android 12) however it also seems to work now (at least the NFC Discovery without crash and Runtime exception). 
I can see my wallet address after my nfc card connected and paired successfully without Exception, but everything else seems to be dummy implemented at this point.

So at least with the PreRelease apk you sent me the Runntime Exception is gone and no crash happens. 
If you need a DebugLog from my device, please ask me again.

Nevertheless it sometimes takes me about 20 tries to get the card connected with the current discovery process, and another time it works instantly. Seems to be another secondary problem.

Is there an ETA of the new RC implementing this fix?

**Comment by bitgamma** [comment]
I have finalized the two PRs (#14997 and #15000) and now the "APDU response must be at least 2 bytes" exception is also handled as IOException as it should.

**Comment by churik** [comment]
@localh0rs7 thank you for reporting the issue!
It is fixed in the upcoming release 1.20.3.
